### DISCUSSION TITLE
[Bug?] Inconsistent types when watching a field array

### DISCUSSION THREAD

**Discussion by slugmandrew** [react-hook-form/react-hook-form#6565] [Q&A]
I am trying to implement a variable size list of complex objects, so am using a field array. I also need to use `react-select`, so am using `Controller`.

When I `watch()` the field array, and then get the values back, I am seeing some very strange behaviour (or perhaps I just don't understand properly).

### Codesandbox

Here is the CSB for this problem, just open one of the select menus to trigger the logging.

https://codesandbox.io/s/react-hook-form-field-array-incorrect-types-yxfbk?file=/src/index.tsx

I will now give some more details to explain my issue...

### Problem

Given the following snippet, I would expect the type of `fieldValues1` and `fieldValues2` to both be `string[]`, however `fieldValues1` is _actually_  an `object[]`.

```
const fieldValues1 = watchFields.map((field) => field.keyDef); // map to string
const fieldValues2 = fields.map((field) => field.keyDef); // map to string

console.log("fieldValues1", fieldValues1);
console.log("fieldValues2", fieldValues2);

```

### Output

![image](https://user-images.githubusercontent.com/153625/133609543-4f342285-bf5d-4336-842b-b2624520a396.png)

Strangely, the watched fields are being mapped into a string[], but although the TypeScript compiler thinks it's a string[], it's actually and array of `{ label: string, value: string }`.

**As you can see, even codesandbox thinks this is an array of strings!**

![image](https://user-images.githubusercontent.com/153625/133610404-07d2f6e1-9ad8-49c9-8a6e-1941a96e8f00.png)

I obviously can't use `fields` instead of `watchFields` as I need the current values (each select depends on the others).

I am getting this in my project also, so it's not just CodeSandbox. What is going on? 😧

Thank you for your time 💙

EDIT: Forgot to say I am using react-hook-form v6.

**Comment by bluebill1049** [comment]
> Given the following snippet, I would expect the type of fieldValues1 and fieldValues2 to both be string[], however fieldValues1 is actually an object[].

could you provide a simple codesandbox for this issue above (remove all the side effects if possible)? There is quite a lot of stuff happen in your existing codesandbox.

**Reply by slugmandrew** [reply]
Sorry about that! I was trying to use my existing data but keep the logic fairly simple (there was only one side-effect) but I get what you mean.

I've simplified it down to its essence here: https://codesandbox.io/s/react-hook-form-field-array-incorrect-types-simplified-yxfbk?file=/src/index.tsx Just open any of the select boxes and you will see the behaviour I mentioned before.

One other thing I have noticed is that **this only happens when the fields are named correctly** - if I change line 89 to 
```name={`${fieldArrayName}x[${index}].username`}``` (note the additional `x`) then it works, though gives the default values of course.

Presumably the behaviour inside `watch` is different depending on whether the values can be successfully obtained or not?

Happy to try to help create a fix if this is a 🐛

**Reply by bluebill1049** [reply]
thanks for the updated codesandbox. ahah the `[]` i reckon it probably confused to an array input compare with a flat one. Any chance you could avoid those special characters? especially those `[ ] .`

**Reply by slugmandrew** [reply]
I'm a bit confused now. I followed the documentation where it says:

> Name of the field. Important: make sure name is in object shape: **name=test[index].name** as we don't support flat arrays.

![image](https://user-images.githubusercontent.com/153625/134407913-9d9364d6-4341-4956-a969-67fb5fda701e.png)

I actually do want this as an array of `Person`. I've got much closer to what I actually want, but still using the same name structure. The only thing missing now is being able to set the label for the `Select`. If you check the [updated CSB](https://codesandbox.io/s/react-hook-form-field-array-incorrect-types-simplified-yxfbk?file=/src/index.tsx) you can see it has the initial values set correctly (`person.username`), and when you change the value everything works as expected, but I can't work out a way to set the initial (I guess default) label, which I want to be `person.name`.

Sorry to be a pain, promise I'll help out as much as I can :)

**Reply by bluebill1049** [reply]
not a pain at all! where is that example? i need to fix it.

**Reply by bluebill1049** [reply]
If you want an array, dot synax is what you after

eg: `test.0.value`

**Reply by slugmandrew** [reply]
> not a pain at all! where is that example? i need to fix it.

It's in the v6 docs: https://react-hook-form.com/v6/api#useFieldArray

Still confused though - There are examples of this kind of naming throughout the docs, such as here: https://react-hook-form.com/advanced-usage#FieldArrays (Register tab).

Is it that the custom hook uses `test.0.value` but if we are creating a 'field array' type system, we use `test[0].value`?

**Reply by slugmandrew** [reply]
FYI I think I have achieved what I want, but using syntax like `test.0` then all of the fields from the object are mapped to that array index. Does that sounds like a sensible usage?

**Reply by bluebill1049** [reply]
oh! you are using v6... sorry for the confusion (not pay attention to the syntax in the CSB) I thought you are on the latest. is any chance to upgrade to the latest v7? for useFieldArray I have to say v7 is a lot better with internal design, performance, and most importantly less unexpected behavior.

**Reply by slugmandrew** [reply]
Ahh yeah that explains it. We're already using v6 in production so was kinda waiting til v7 was stable (started this project in March).

I'll give v7 a go then!

Thanks as always.

**Reply by bluebill1049** [reply]
Thanks for your patient and understanding, I honestly wish everyone in the community act the same. 🙏

**Reply by bluebill1049** [reply]
by the way, v7 is definitely stable as we speak. 💪

**Reply by slugmandrew** [reply]
There is definitely some funky behaviour going on in v6, particularly relating to the type of the `watch(fields)`. 

It is probably not a bug but a 'quirk' of the library. ;) It makes a bit more sense now but maybe you can explain a little how it works...

Consider:

```
type MyIntersectionType = TypeA & TypeB
type MyFormType = {
  arrayOfThings: Array<MyIntersectionType>
}

const form = useForm<MyFormType>()
const { fields, append, remove } = useFieldArray<TypeA>({ form.control, name: "arrayOfThings" }) // type = TypeA[]
const watchFields = watch("arrayOfThings") // type = MyIntersectionType[]

```

In the above example I would have expected `watchFields` to be an array of `TypeA`, but it is of `MyIntersectionType`, due to inspecting the type of `MyFormType.arrayOfThings`, I think.

So it seems the generic type argument passed to `useFieldArray<TypeA>` is not used by this `watch` at all.

This is what was confusing me, because `fields.map()` gives a set of `TypeA`, but then when I set a new value in `OnChange`, the values in watch end up like:

```
[
MyIntersectionType,
MyIntersectionType,
TypeA // new value
]
```

I have 2 questions:

1) Should the type of the field array have to match the type in the form exactly?
2) Is there a way to parameterise `watch`, such as `watch<TypeA>("name")`? (or can we make that in v8? 🙂)

**Reply by bluebill1049** [reply]
> There is definitely some funky behaviour going on in v6, particularly relating to the type of the watch(fields).

I think so. There are few changes that were made to fulfill the gap between re-render especially with field array at v6, but got a proper solution at v7.

> Should the type of the field array have to match the type in the form exactly?

yes

> Is there a way to parameterise watch, such as watch<TypeA>("name")? (or can we make that in v8? 🙂)

what do you mean? could you provide some more contextual examples?

**Reply by slugmandrew** [reply]
`watch<T>(string: name)` instead of `watch(string: name)`, so you can tell it what type it is watching.

**Reply by bluebill1049** [reply]
oh perhaps we can do something similar: https://github.com/react-hook-form/react-hook-form/pull/6562 the only risky is you are basically skipping the type check.
